ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceManager.java

```diff
@@ -80,66 +86,97 @@ public void run() {
             timer.schedule(refreshIngestionResourceValuesTask, 0, REFRESH_INGESTION_RESOURCES_PERIOD);
 
         } catch (Exception e) {
-            log.error(String.format("Error in initializing ResourceManager: %s.", e.getMessage()), e);
+            log.error("Error in initializing ResourceManager: {}.", e.getMessage(), e);
             throw e;
         }
     }
 
-    public void clean() {
+    void clean() {
         ingestionResources.clear();
     }
 
-    public String getKustoIdentityToken() throws Exception {
-        if (identityToken == null) {
-            refreshIngestionAuthToken();
-            if (identityToken == null) {
-                throw new Exception("Unable to get Identity token");
+    String getIngestionResource(ResourceTypes resourceType) throws Exception {
+        if (!ingestionResources.containsKey(resourceType)) {
+            // When the value is not available, we need to get the tokens from Kusto (refresh):
+            refreshIngestionResources();
+            try{
+                // If the write lock is locked, then the read will wait here.
+                // In other words if the refresh is running yet, then wait until it ends:
+                ingestionResourcesLock.readLock().lock();
+                if (!ingestionResources.containsKey(resourceType)) {
+                    throw new Exception("Unable to get ingestion resources for this type: " + resourceType.getName());
+                }
+            } finally {
+                ingestionResourcesLock.readLock().unlock();
```

Comment:
Can it be that the lock failed, and than we'll get here another error?

Reply:
if the lock fails due to error, it will throw an exception to the caller of this method.
if the lock is already locked, it will wait here until it released.
(or I didn't understand the question :)